aws/resource_aws_budgets_budget_test.go

```diff
@@ -0,0 +1,408 @@
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_unit", configBasicUpdate.TimeUnit),
+					testBudgetExists(configBasicUpdate, testAccProvider),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSBudget_prefix(t *testing.T) {
+	name := "test-budget-"
+	configBasicDefaults := newBudgetTestConfigDefaults(name)
+	configBasicUpdate := newBudgetTestConfigUpdate(name)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		CheckDestroy: func(s *terraform.State) error {
+			return testCheckBudgetDestroy(name, testAccProvider)
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testBudgetHCLPrefixUseDefaults(configBasicDefaults),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckNoResourceAttr("aws_budgets_budget.foo", "account_id"),
+					resource.TestMatchResourceAttr("aws_budgets_budget.foo", "name_prefix", regexp.MustCompile(configBasicDefaults.BudgetName)),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "budget_type", configBasicDefaults.BudgetType),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "limit_amount", configBasicDefaults.LimitAmount),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "limit_unit", configBasicDefaults.LimitUnit),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_period_start", configBasicDefaults.TimePeriodStart),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_period_end", configBasicDefaults.TimePeriodEnd),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_unit", configBasicDefaults.TimeUnit),
+					testBudgetExists(configBasicDefaults, testAccProvider),
+				),
+			},
+
+			{
+				Config: testBudgetHCLPrefix(configBasicUpdate),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckNoResourceAttr("aws_budgets_budget.foo", "account_id"),
+					resource.TestMatchResourceAttr("aws_budgets_budget.foo", "name_prefix", regexp.MustCompile(configBasicUpdate.BudgetName)),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "budget_type", configBasicUpdate.BudgetType),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "limit_amount", configBasicUpdate.LimitAmount),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "limit_unit", configBasicUpdate.LimitUnit),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_period_start", configBasicUpdate.TimePeriodStart),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_period_end", configBasicUpdate.TimePeriodEnd),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_unit", configBasicUpdate.TimeUnit),
+					testBudgetExists(configBasicUpdate, testAccProvider),
+				),
+			},
+		},
+	})
+}
+
+func testBudgetExists(config budgetTestConfig, provider *schema.Provider) resource.TestCheckFunc {
```

Comment:
Generally we separate out existence tests and attribute tests.

The signature for this should be `testAccCheckAWSBudgetsBudgetExists(resourceName string, budget *budget.Budget) resource.TestCheckFunc`

Where the provider is already available via an already initialized `testAccProvider` variable, so `testAccProvider.Meta().(*AWSClient).budgetconn` works automatically.

And the budget is made available for other attribute check functions via `*budget = *b.Budget` before returning.

Reply:
my thought in putting the provider as an argument in these functions was to avoid dependencies on those global vars. I guess since they are there to be used and are standard, it makes sense to conform.

with regard to your last comment of setting `*budget = *b.Budget` im not sure i follow what you're suggesting. Could you expand that thought a bit?